File: apps/android/app/src/main/java/com/apostolos/tv/data/sync/UserSyncManager.kt
package com.apostolos.tv.data.sync

import com.apostolos.tv.data.CategoryVisibilityStore
import com.apostolos.tv.data.CredentialsStore
import com.apostolos.tv.data.FavoritesStore
import com.apostolos.tv.data.WatchHistoryStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class UserSyncManager(
    private val authApi: SupabaseAuthApi,
    private val cloudSync: CloudSyncRepository,
    private val accountSessionStore: AccountSessionStore,
    private val credentialsStore: CredentialsStore,
    private val watchHistoryStore: WatchHistoryStore,
    private val favoritesStore: FavoritesStore,
    private val categoryVisibilityStore: CategoryVisibilityStore,
) {
    data class SyncStatus(
        val lastSyncedAt: Long? = null,
        val isSyncing: Boolean = false,
        val lastError: String? = null,
    )

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val _syncStatus = MutableStateFlow(SyncStatus())
    val syncStatus: StateFlow<SyncStatus> = _syncStatus.asStateFlow()
    private var pushJob: Job? = null
    private var suppressPush = false

    fun startAutoSync() {
        scope.launch {
            combine(
                credentialsStore.credentialsFlow,
                watchHistoryStore.entries,
                favoritesStore.items,
                categoryVisibilityStore.state,
            ) { _, _, _, _ -> Unit }
                .collect {
                    if (!suppressPush && accountSessionStore.session.value != null) {
                        schedulePush()
                    }
                }
        }
    }

    suspend fun resolveSession(): AccountSession? = withContext(Dispatchers.IO) {
        val current = accountSessionStore.session.value ?: return@withContext null
        if (!current.isExpired()) return@withContext current

        runCatching {
            val refreshed = authApi.refreshSession(current)
            accountSessionStore.save(refreshed)
            refreshed
        }.getOrElse {
            accountSessionStore.clear()
            null
        }
    }

    suspend fun syncAfterAccountLogin(session: AccountSession): UserSyncPayload = withContext(Dispatchers.IO) {
        accountSessionStore.save(session)
        val activeSession = resolveSession() ?: session
        val remote = runCatching { cloudSync.pull(activeSession) }.getOrNull()

        suppressPush = true
        try {
            if (remote != null) {
                applyPayload(remote)
                _syncStatus.update {
                    it.copy(lastSyncedAt = System.currentTimeMillis(), lastError = null)
                }
                remote
            } else {
                val local = collectPayload()
                cloudSync.push(activeSession, local)
                _syncStatus.update {
                    it.copy(lastSyncedAt = System.currentTimeMillis(), lastError = null)
                }
                local
            }
        } finally {
            suppressPush = false
        }
    }

    suspend fun pushNow() = withContext(Dispatchers.IO) {
        _syncStatus.update { it.copy(isSyncing = true, lastError = null) }
        try {
            val session = resolveSession() ?: throw CloudSyncException("Δεν είσαι συνδεδεμένος.")
            cloudSync.push(session, collectPayload())
            _syncStatus.update {
                it.copy(isSyncing = false, lastSyncedAt = System.currentTimeMillis(), lastError = null)
            }
        } catch (error: Exception) {
            _syncStatus.update { it.copy(isSyncing = false, lastError = error.message) }
            throw error
        }
    }

    suspend fun pullFromCloud() = withContext(Dispatchers.IO) {
        _syncStatus.update { it.copy(isSyncing = true, lastError = null) }
        try {
            val session = resolveSession() ?: throw CloudSyncException("Δεν είσαι συνδεδεμένος.")
            val remote = cloudSync.pull(session) ?: throw CloudSyncException("Δεν υπάρχουν δεδομένα στο cloud.")
            applyPayload(remote)
            _syncStatus.update {
                it.copy(isSyncing = false, lastSyncedAt = System.currentTimeMillis(), lastError = null)
            }
        } catch (error: Exception) {
            _syncStatus.update { it.copy(isSyncing = false, lastError = error.message) }
            throw error
        }
    }

    suspend fun changePassword(newPassword: String) = withContext(Dispatchers.IO) {
        val session = resolveSession() ?: throw SupabaseAuthException("Δεν είσαι συνδεδεμένος.")
        authApi.updatePassword(session, newPassword)
    }

    suspend fun applyPayload(payload: UserSyncPayload) {
        suppressPush = true
        try {
            watchHistoryStore.replaceAll(payload.watchHistory)
            favoritesStore.replaceAll(payload.favorites)
            categoryVisibilityStore.replaceAll(payload.categoryVisibility)
            credentialsStore.replaceAll(payload.playlists)
        } finally {
            suppressPush = false
        }
    }

    suspend fun clearLocalData() {
        suppressPush = true
        try {
            credentialsStore.clear()
            watchHistoryStore.clear()
            favoritesStore.clear()
            categoryVisibilityStore.clear()
        } finally {
            suppressPush = false
        }
    }

    suspend fun logout() = withContext(Dispatchers.IO) {
        runCatching { pushNow() }
        accountSessionStore.clear()
        clearLocalData()
    }

    fun collectPayload(): UserSyncPayload = UserSyncPayload(
        playlists = credentialsStore.playlistsState.value,
        watchHistory = watchHistoryStore.entries.value,
        favorites = favoritesStore.items.value,
        categoryVisibility = categoryVisibilityStore.state.value,
    )

    private fun schedulePush() {
        pushJob?.cancel()
        pushJob = scope.launch {
            delay(PUSH_DEBOUNCE_MS)
            _syncStatus.update { it.copy(isSyncing = true, lastError = null) }
            runCatching { pushNow() }
        }
    }

    companion object {
        private const val PUSH_DEBOUNCE_MS = 2_500L
    }
}

File: apps/tizen-tv/src/sync/userSyncManager.ts
import type { XtreamCredentials } from "@tv/xtream-core";
import { getAppConfig, isSupabaseConfigured } from "../config.js";
import { categoryVisibilityStore } from "../data/categoryVisibility.js";
import { favoritesStore } from "../data/favoritesStore.js";
import { watchHistoryStore, type WatchEntry } from "../data/watchHistory.js";
import { pullSync, pushSync, refreshSession, signIn, signUp } from "./cloudSync.js";
import {
  activeCredentials,
  createPlaylist,
  isSessionExpired,
  type AccountSession,
  type PlaylistsState,
  type UserSyncPayload,
} from "./types.js";
import { clearSession, loadSession, saveSession } from "./sessionStore.js";

const PLAYLISTS_KEY = "tv_playlists";
const PUSH_DEBOUNCE_MS = 800;

export interface SyncStatus {
  lastSyncedAt: number | null;
  lastError: string | null;
  isSyncing: boolean;
}

export class UserSyncManager {
  private syncStatus: SyncStatus = {
    lastSyncedAt: null,
    lastError: null,
    isSyncing: false,
  };

  private pushTimer: ReturnType<typeof setTimeout> | null = null;
  private suppressPush = false;

  getStatus(): SyncStatus {
    return { ...this.syncStatus };
  }

  loadPlaylists(): PlaylistsState {
    try {
      const raw = localStorage.getItem(PLAYLISTS_KEY);
      if (!raw) return { playlists: [], activePlaylistId: null };
      return JSON.parse(raw) as PlaylistsState;
    } catch {
      return { playlists: [], activePlaylistId: null };
    }
  }

  savePlaylists(state: PlaylistsState): void {
    localStorage.setItem(PLAYLISTS_KEY, JSON.stringify(state));
  }

  getCredentials(): XtreamCredentials | null {
    return activeCredentials(this.loadPlaylists());
  }

  saveCredentials(credentials: XtreamCredentials, name = "Κύρια playlist"): void {
    const current = this.loadPlaylists();
    const existing = current.playlists[0];
    const playlist = existing
      ? { ...existing, ...credentials, name: name.trim() || existing.name }
      : createPlaylist(name, credentials);
    this.savePlaylists({
      playlists: [playlist],
      activePlaylistId: playlist.id,
    });
  }

  collectPayload(): UserSyncPayload {
    return {
      playlists: this.loadPlaylists(),
      watchHistory: watchHistoryStore.getAll(),
      favorites: favoritesStore.getAll(),
      categoryVisibility: {
        hiddenLive: categoryVisibilityStore.hiddenLive,
        hiddenMovies: categoryVisibilityStore.hiddenMovies,
        hiddenSeries: categoryVisibilityStore.hiddenSeries,
      },
    };
  }

  applyPayload(payload: UserSyncPayload): void {
    this.suppressPush = true;
    try {
      this.savePlaylists(payload.playlists);
      watchHistoryStore.replaceAll(payload.watchHistory);
      favoritesStore.replaceAll(payload.favorites);
      categoryVisibilityStore.replaceAll(payload.categoryVisibility);
    } finally {
      this.suppressPush = false;
    }
  }

  schedulePush(): void {
    if (this.suppressPush) return;
    if (this.pushTimer) clearTimeout(this.pushTimer);
    this.pushTimer = setTimeout(() => {
      void this.pushNow().catch(() => {
        // ignore background sync errors
      });
    }, PUSH_DEBOUNCE_MS);
  }

  onWatchHistoryChanged(): void {
    this.onDataChanged();
  }

  onDataChanged(): void {
    if (loadSession()) this.schedulePush();
  }

  async resolveSession(): Promise<AccountSession | null> {
    if (!isSupabaseConfigured(getAppConfig())) return null;
    const current = loadSession();
    if (!current) return null;
    if (!isSessionExpired(current)) return current;
    try {
      const refreshed = await refreshSession(current);
      saveSession(refreshed);
      return refreshed;
    } catch {
      clearSession();
      return null;
    }
  }

  async syncAfterAccountLogin(session: AccountSession): Promise<XtreamCredentials | null> {
    saveSession(session);
    this.syncStatus.isSyncing = true;
    try {
      const activeSession = (await this.resolveSession()) ?? session;
      const remote = await pullSync(activeSession);
      if (remote) {
        this.applyPayload(remote);
        this.syncStatus.lastSyncedAt = Date.now();
        this.syncStatus.lastError = null;
        return activeCredentials(remote.playlists);
      }
      const local = this.collectPayload();
      await pushSync(activeSession, local);
      this.syncStatus.lastSyncedAt = Date.now();
      this.syncStatus.lastError = null;
      return activeCredentials(local.playlists);
    } catch (error) {
      this.syncStatus.lastError =
        error instanceof Error ? error.message : "Sync failed.";
      throw error;
    } finally {
      this.syncStatus.isSyncing = false;
    }
  }

  async signInAccount(email: string, password: string, register = false): Promise<XtreamCredentials | null> {
    const session = register ? await signUp(email, password) : await signIn(email, password);
    return this.syncAfterAccountLogin(session);
  }

  async pullNow(): Promise<void> {
    const session = await this.resolveSession();
    if (!session) throw new Error("Δεν είσαι συνδεδεμένος.");
    this.syncStatus.isSyncing = true;
    try {
      const remote = await pullSync(session);
      if (remote) {
        this.applyPayload(remote);
      }
      this.syncStatus.lastSyncedAt = Date.now();
      this.syncStatus.lastError = null;
    } catch (error) {
      this.syncStatus.lastError = error instanceof Error ? error.message : "Sync failed.";
      throw error;
    } finally {
      this.syncStatus.isSyncing = false;
    }
  }

  async pushNow(): Promise<void> {
    const session = await this.resolveSession();
    if (!session) throw new Error("Δεν είσαι συνδεδεμένος.");
    await pushSync(session, this.collectPayload());
    this.syncStatus.lastSyncedAt = Date.now();
  }

  async logout(): Promise<void> {
    try {
      await this.pushNow();
    } catch {
      // ignore push errors on logout
    }
    clearSession();
    localStorage.removeItem(PLAYLISTS_KEY);
    watchHistoryStore.replaceAll([]);
    favoritesStore.replaceAll([]);
    categoryVisibilityStore.replaceAll({ hiddenLive: [], hiddenMovies: [], hiddenSeries: [] });
  }
}

export const userSyncManager = new UserSyncManager();

export type { WatchEntry };
